Make systemctl list-unit-files output more useful

keszybz opened the pull request:
For #15205.

poettering commented on src/shared/unit-file.c:
```diff
@@ -329,8 +329,9 @@ int unit_file_build_name_map(
 
-                                if (!path_is_absolute(target)) {
-                                        target_abs = path_join(*dir, target);
+                                const bool is_abs = path_is_absolute(target);
```
what does the "const" du here? it's very confusing suddently using that for stuff that isn't a ptr...

keszybz replied:
I'm just saying that this a variable that will not be modified later on. A hint for the reader.

keszybz replied:
I can remove it if you think it's absolutely wrong...

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -5860,7 +5860,7 @@ static int show(int argc, char *argv[], void *userdata) {
         if (show_mode == SYSTEMCTL_SHOW_HELP && argc <= 1)
                 return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
-                                       "This command expects one or more unit names. Did you mean --help?");
+                                       "'help' is not supported as a command verb. Did you mean --help?");
```
hmm, uh? i don't get this? aren't you confusing the user here a bit? I mean `systemctl help systemd-journald.service` is a thing, but with the new msg you suggest otherwise?

keszybz replied:
Oh, OK. I misunderstood this completely. Will rework.

poettering commented on src/shared/verbs.c:
```diff
@@ -45,10 +45,23 @@ bool running_in_chroot_or_offline(void) {
 }
 
+const Verb* verbs_find_verb(const char *name, const Verb verbs[]) {
+        for (unsigned i = 0;; i++) {
```
we iterate through memory here, please use `size_t`

keszybz replied:
I was just moving existing code... Will fix.

poettering commented on src/shared/verbs.c:
```diff
@@ -45,10 +45,23 @@ bool running_in_chroot_or_offline(void) {
+        for (unsigned i = 0;; i++) {
+                /* At the end of the list? */
+                if (!verbs[i].dispatch)
+                        break;
```
why not move this test into the `for()` condition above?

poettering commented on src/shared/verbs.c:
```diff
@@ -45,10 +45,23 @@ bool running_in_chroot_or_offline(void) {
+                        break;
+
+                if (streq_ptr(name, verbs[i].verb) ||
+                    (!name && verbs[i].flags & VERB_DEFAULT))
```
hmm, && and & in one expresion is a bit hard to read. consider using () or FLAGS_SET()

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -9212,6 +9212,12 @@ static int systemctl_main(int argc, char *argv[]) {
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Verb '%s' cannot be used with --root=.",
+                                       argv[optind] ?: verb->verb);
+
```
hmm, another option would be to simply set the `$SYSTEMD_OFFLINE` env var when `--root=` is specified. The right thing happens then anyway...

I think setting the env var is slightly nicer. dunno?

keszybz replied:
The opposite for me ;) think the SYSTEMD_OFFLINE thing is very vague, because for most operations, it makes no sense to do nothing when offline. For example, `systemctl log-level` returns "", which is not one of the values in the enumeration and will most likely confuse any consumer. I think it'd be more appropriate to refuse operation. But handling compatiblity will existing users of this is hard, and I didn't want to open that can of worms. So I very much prefer to not use the variable internally if that's OK.

poettering replied:
the reason i preferred the $SYSTEMD_OFFLINE approach was that we now iterate through the verbs list twice.., once early, beforehand and once for real. hat's not sexy...

Maybe another option is to add a flags param to dispatch_verb forcing offline mode detection, which we could set here depending on arg_root. Then we'd iterate trough everything once only, and the errors about online/offline would be generated at a single place only.

poettering commented on src/shared/install.h:
```diff
@@ -189,7 +189,26 @@ void unit_file_dump_changes(int r, const char *verb, const UnitFileChange *chang
+        bool initialized;
+} Presets;
+
+void presets_freep(Presets *p);
```
hmm, everything in this file is prefixed `unit_file_`. I think this stuff should be preset too if its exported now.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -1480,16 +1480,20 @@ static int output_unit_file_list(const UnitFileList *units, unsigned c) {
-                        unit_preset_str = "enabled";
-                        on_preset_color = underline ? ansi_highlight_green_underline() : ansi_highlight_green();
+                if (IN_SET(u->state, UNIT_FILE_ALIAS, UNIT_FILE_STATIC))
+                        unit_preset_str = "-"; /* Don't show preset state for aliases, it'll only confuse users. */
```
hmm, please insert a TABLE_EMPTY cell instead. That's useful for alternative ouputs (ie. JSON), and adds implicit colorouing...

i.e. use the table_add_many() below just for the first two fields. and then add the preset field with a separate call that either adds TABLE_EMPTY or TABLE_STRING depending if we have a preset or not.

poettering commented:
hmm, the "systemctl status" output should suppress the preset output on alias/static too

keszybz commented:
Updated. I implemented the comments, excepts as noted above.
I added "generated" and "transient" units to the skip list, and `status` now uses the same logic.

poettering commented on src/shared/install.h:
```diff
@@ -189,7 +189,26 @@ void unit_file_dump_changes(int r, const char *verb, const UnitFileChange *chang
+        PresetRule *rules;
+        size_t n_rules;
+        bool initialized;
+} Presets;
```
i think the three types should be prefixed too

keszybz replied:
I moved them back to install.c instead. No need to expose too much stuff.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -1441,6 +1441,16 @@ static bool output_show_unit_file(const UnitFileList *u, char **states, char **p
+                       UNIT_FILE_TRANSIENT);
+}
+
+
```
spurious double newline

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -4223,10 +4242,17 @@ static void print_status_info(
+
+        } else if (path && !isempty(i->unit_file_state))
                 printf("     Loaded: %s%s%s (%s; %s)\n",
                        on, strna(i->load_state), off, path, i->unit_file_state);
```
you are filtering the preset stuff twice now. the final branch was supposed to be the filtering branch...

i.e. the check for STR_IN_SET(i->unit_file_state, "generated", "transient") is to be subsumed by the show_preset_for_state() check...

keszybz replied:
Duh.

poettering commented:
did another review round. the comments about "systemct status" matter, the others are just nitpicking...

keszybz commented:
Updated with the suggested changes.
